charmbits/httpservice/server.go

```diff
@@ -0,0 +1,439 @@
+	}
+	srv.tomb.Go(func() error {
+		srv.tomb.Kill(rpcCmd.Wait())
+		return nil
+	})
+	srv.tomb.Go(func() error {
+		<-srv.tomb.Dying()
+		// The command has been killed. Stop the RPC listener
+		// and close any current listeners to cause the handlers
+		// to terminate.
+		rpcCmd.Kill()
+		err := rpcCmd.Wait()
+		if err != nil {
+			err = errgo.Notef(err, "local RPC server")
+		}
+
+		srv.mu.Lock()
+		defer srv.mu.Unlock()
+		srv.closeResources()
+		return err
+	})
+	return srv, nil
+}
+
+// Kill implements gocharm.Command.Kill.
+func (srv *server) Kill() {
+	srv.tomb.Kill(nil)
+}
+
+// Wait implements gocharm.Command.Wait.
+func (srv *server) Wait() error {
+	return srv.tomb.Wait()
+}
+
+// Feedback is an implementation detail, exposed only because net/rpc requires it.
+type Feedback struct {
+	Warnings []string
+}
+
+// show shows all the warnings.
+func (fb *Feedback) show() {
+	for _, w := range fb.Warnings {
+		fmt.Fprintf(os.Stderr, "warning: %v\n", w)
+	}
+}
+
+// addError adds the given error to the list of warnings
+// that will be reported as the result of an RPC call.
+func (fb *Feedback) addError(err error) {
+	fb.Warnings = append(fb.Warnings, err.Error())
+}
+
+// Changed implements an RPC server method that is called
+// when a value changes in the charm.
+func (srv *server) Set(p *ServerState, response *Feedback) error {
+	srv.mu.Lock()
+	defer srv.mu.Unlock()
+	srv.set(*p, response)
+	srv.state = *p
+	// Save the server state so that if it's started or the
+	// instance is rebooted, we can carry on as before.
+	if err := srv.saveState(srv.state); err != nil {
+		return errgo.Notef(err, "cannot save server state")
+	}
+	return nil
+}
+
+func (srv *server) serveHTTP(port int, h http.Handler) error {
+	addr := ":" + strconv.Itoa(port)
+	listener, err := net.Listen("tcp", addr)
+	if err != nil {
+		return errgo.Newf("cannot listen on %s: %v", addr, err)
+	}
+	server := &http.Server{
+		Addr:    addr,
+		Handler: h,
+	}
+	srv.httpListener = newHandlerListener(server, listener)
+	return nil
+}
+
+func (srv *server) serveHTTPS(port int, certPEM string, h http.Handler) error {
+	certPEMBytes := []byte(certPEM)
+	cert, err := tls.X509KeyPair(certPEMBytes, certPEMBytes)
+	if err != nil {
+		return errgo.Newf("cannot parse certificate: %v", err)
+	}
+	config := &tls.Config{
+		NextProtos:   []string{"http/1.1"},
+		Certificates: []tls.Certificate{cert},
+	}
+	addr := ":" + strconv.Itoa(port)
+	listener, err := net.Listen("tcp", addr)
+	if err != nil {
+		return errgo.Newf("cannot listen on %s: %v", addr, err)
+	}
+	tlsListener := tls.NewListener(
+		tcpKeepAliveListener{listener.(*net.TCPListener)},
+		config,
+	)
+	server := &http.Server{
+		Addr:    addr,
+		Handler: h,
+	}
+	srv.httpsListener = newHandlerListener(server, tlsListener)
+	return nil
+}
+
+func (srv *server) set(state ServerState, fb *Feedback) {
+	srv.set0(state, fb)
+	log.Printf("httpservice set -> %q", fb.Warnings)
```

Comment:
So that I can show the warnings regardless of when it returns. I've added the show-warnings logic to the caller instead and ditched set0.